snuba/snuba_migrations/group_attributes/0001_group_attributes.py

```diff
@@ -0,0 +1,71 @@
+from typing import List, Sequence
+
+from snuba.clickhouse.columns import Column, DateTime, UInt
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.migrations.operations import OperationTarget, SqlOperation
+
+columns: List[Column[Modifiers]] = [
+    Column("project_id", UInt(64)),
+    Column("group_id", UInt(64)),
+    # Group
+    Column("group_status", UInt(8)),
+    Column("group_substatus", UInt(8, Modifiers(nullable=True))),
+    Column("group_first_seen", DateTime()),
+    Column("group_num_comments", UInt(64)),
+    # GroupAssignee
+    Column("assignee_user_id", UInt(64, Modifiers(nullable=True))),
+    Column("assignee_team_id", UInt(64, Modifiers(nullable=True))),
+    # GroupOwner
+    Column("owner_suspect_commit_user_id", UInt(64, Modifiers(nullable=True))),
+    Column("owner_ownership_rule_user_id", UInt(64, Modifiers(nullable=True))),
+    Column("owner_ownership_rule_team_id", UInt(64, Modifiers(nullable=True))),
+    Column("owner_codeowners_user_id", UInt(64, Modifiers(nullable=True))),
+    Column("owner_codeowners_team_id", UInt(64, Modifiers(nullable=True))),
+    Column("message_timestamp", DateTime()),
+    Column("partition", UInt(16)),
+    Column("offset", UInt(64)),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.GROUP_ATTRIBUTES,
+                table_name="group_attributes_local",
+                columns=columns,
+                engine=table_engines.ReplacingMergeTree(
+                    order_by="(project_id, group_id",
+                    settings={"index_granularity": "8192"},
+                    storage_set=StorageSetKey.GROUP_ATTRIBUTES,
```

Comment:
I omitted certain common table engine parameters like `version_column`, `partition_by`, `sample_by`, and `ttl`.

Since this data is a mirror of the postgres Groups tables, I think we'll need to retain the data indefinitely.

Reply:
Don't we need the ability to delete the group and null out the data though, for which a version column may be helpful? If a column is deleted from postgres, we should remove it here too?

Reply:
Oh, right. Forgot about when we actually delete the Group row. In the case of `GroupAssignee` and `GroupOwner` rows getting deleted, I was planning to simply send a new snapshot with the appropriate column values nulled out. But that doesn't really work when the Group is totally deleted. Will include a deleted column to account for this.

On a somewhat related note, does clickhouse store timestamp metadata for every row insert? I'm wondering how the replacing merge tree knows which duplicate rows are 'latest' without us having to specify a timestamp to impose ordering for de-duping:

```
CREATE TABLE test_table(`key` Int64, `time` DateTime) ENGINE = ReplacingMergeTree() ORDER BY key;
INSERT INTO test_table Values (1, now());
INSERT INTO test_table Values (1, now());
INSERT INTO test_table Values (1, now());
INSERT INTO test_table Values (1, yesterday());

----

select * from test_table

┌─key─┬────────────────time─┐
│   1 │ 2023-07-10 00:00:00 │
└─────┴─────────────────────┘
┌─key─┬────────────────time─┐
│   1 │ 2023-07-11 23:27:48 │
└─────┴─────────────────────┘
┌─key─┬────────────────time─┐
│   1 │ 2023-07-11 23:27:47 │
└─────┴─────────────────────┘
┌─key─┬────────────────time─┐
│   1 │ 2023-07-11 23:27:45 │
└─────┴─────────────────────┘

----

select * from test_table final

┌─key─┬────────────────time─┐
│   1 │ 2023-07-10 00:00:00 │
└─────┴─────────────────────┘
```

notice the first three inserts results in dupe rows the increasing `time` value, and the final insert is out of order, but when we apply `FINAL` clickhouse knows that the row with the `time=yesterday()` is the 'latest'.

Reply:
Since you didn't specify a version column, ClickHouse defaults to keeping the last inserted row. See https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replacingmergetree#ver